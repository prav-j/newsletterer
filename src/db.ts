import { Sequelize as Sequelize } from "sequelize-typescript";

import configLoader from "./config";
import * as path from "path";
import { Transaction } from "sequelize";

const config = configLoader();

let sequelize: Sequelize;

export function getDBInstance(): Sequelize {
  if (!sequelize) {
    const {db: connectionUri} = config
    sequelize = new Sequelize(connectionUri, {
      logging: false,
      pool: {
        min: 1,
        max: 20,
        idle: 10000
      }
    });
    sequelize.addModels([path.join(__dirname, 'modules/**/*.model.*')])
  }
  return sequelize;
}

export const withTransaction = <T>(callback: (t: Transaction) => Promise<T>): Promise<T> => {
  return getDBInstance().transaction(callback)
}

export async function initializeDB(): Promise<Sequelize> {
  const sequelize = getDBInstance();
  await sequelize.sync({alter: true});
  return sequelize
}

export async function resetDB(): Promise<Sequelize> {
  const sequelize = getDBInstance();
  if (process.env.NODE_ENV !== 'test') {
    console.error('This WILL wipe out all data. Aborting. You may run this only in test mode.')
    return sequelize
  }
  await sequelize.sync({force: true});
  return sequelize
}
