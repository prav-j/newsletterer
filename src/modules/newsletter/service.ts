import User from "../users/User.model";
import { getTopPostsInLastDay } from "../posts/service";
import { UUID } from "../../types/UUID";
import Subreddit from "../subreddit/Subreddit.model";
import * as emailAPI from "../../api/emailService"
import NewsletterSchedule from "./NewsletterSchedule.model";

const createNewsletterForUser = async (user: User) => {
  const states = await Promise.all(user.subreddits.map(subreddit => getTopPostsInLastDay(subreddit)))
  return {
    user: user.name,
    email: user.email,
    subreddits: states
      .filter(state => state.posts.length > 0)
      .map(state => {
        return {
          url: state.link,
          posts: state.posts
        }
      })
  }
}

export const emailNewsletterForUser = async (userId: UUID): Promise<Newsletter | undefined> => {
  const user = await User.findOne({where: {id: userId}, include: [NewsletterSchedule, Subreddit]})
  if (!user || !user.newsletterSchedule.isEnabled) {
    return
  }
  const newsletter = await createNewsletterForUser(user)
  await emailAPI.emailNewsletter(newsletter)
  return newsletter
}

export interface Newsletter {
  user: User['name']
  email: User['email']
  subreddits: {
    url: Subreddit['url'],
    posts: {
      thumbnail?: string,
      title: string
      link: string
    }[]
  }[]
}