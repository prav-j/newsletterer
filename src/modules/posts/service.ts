import fetch from 'node-fetch'
import Subreddit from "../subreddit/Subreddit.model";

import * as xml2js from 'xml2js';

interface RSSFeed {
  feed: {
    logo: [string]
    title: [string]
    subtitle: [string]
    entry: {
      "media:thumbnail"?: {
        $: { url: string }
      }[]
      link: {
        $: { href: string }
      }[]
      title: [string]
    }[]
  }
}

interface SubredditPost {
  thumbnail?: string
  link: string
  title: string
}

interface SubredditState {
  link: string
  logo?: string
  title: string
  subtitle?: string
  posts: SubredditPost[]
}

const POSTS_PER_SUBREDDIT = 3

export const getTopPostsInLastDay = async (subreddit: Subreddit) => {
  const topPostsRssURL = `${subreddit.url}/top/.rss?t=day`
  const result = await fetch(topPostsRssURL)
  const xml = await result.text()
  const feed: RSSFeed = await xml2js.parseStringPromise(xml)
  return getTopPostsFromFeed(subreddit, feed)
}

const getTopPostsFromFeed = (subreddit: Subreddit, {feed}: RSSFeed): SubredditState => {
  const getFirst = (item?: any[]) => item && item[0]
  return {
    link: subreddit.url,
    logo: getFirst(feed.logo),
    title: getFirst(feed.title),
    subtitle: getFirst(feed.subtitle),
    posts: feed.entry.slice(0, POSTS_PER_SUBREDDIT).map(entry => {
      return ({
        thumbnail: getFirst(entry["media:thumbnail"])?.$.url,
        link: entry.link[0].$.href,
        title: getFirst(entry.title)
      });
    })
  }
}