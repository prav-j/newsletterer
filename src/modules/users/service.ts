import User from "./User.model";
import { withTransaction } from "../../db";
import { UUID } from "../../types/UUID";
import { Transaction } from "sequelize";
import NewsletterSchedule from "../newsletter/ScheduledNewsletter.model";

interface CreateOrUpdateUserRequest {
  name: string
  email: string
  isNewsletterEnabled: boolean | undefined,
  newsletterSendTime: string | undefined,
  timezone: string | undefined
}

export async function getUsers() {
  return (await User.findAll()).map(({id, name, email}) => ({id, name, email}));
}

export async function fetchUser(userId: UUID, transaction?: Transaction) {
  return User.findOne({where: {id: userId}, transaction});
}

const extractFromRequest = (request: CreateOrUpdateUserRequest): Partial<User> => {
  return {
    name: request.name,
    email: request.email,
    isNewsletterEnabled: request.isNewsletterEnabled,
    newsletterSendTime: request.newsletterSendTime,
    timezone: request.timezone,
  }
}

export const createUser = async (request: CreateOrUpdateUserRequest) => {
  return await withTransaction(async transaction => {
    const user = await User.create(extractFromRequest(request), {transaction})
    const newsletterSchedule = await NewsletterSchedule.create({userId: user.id}, {transaction})
    if (request.isNewsletterEnabled === false) {
      await newsletterSchedule.cancelScheduled(transaction)
    }
    user.$set('newsletterSchedule', newsletterSchedule)
    return user
  })
}

export async function updateUser(userId: string, request: CreateOrUpdateUserRequest) {
  return withTransaction(async transaction => {
    const user = await fetchUser(userId, transaction)
    if (!user) {
      return null
    }
    return user.update(extractFromRequest(request), {transaction});
  })
}