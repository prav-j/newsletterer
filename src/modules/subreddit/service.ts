import { withTransaction } from "../../db";
import Subreddit from "./Subreddit.model";
import { UUID } from "../../types/UUID";
import * as userService from '../users/service'
import { Transaction, ValidationError } from "sequelize";

export const fetchSubreddit = async (name: string, transaction?: Transaction) => {
  const subreddit = await Subreddit.findOne({where: {name}, transaction})
  if (subreddit) {
    return subreddit
  }
  return null
};

export const subscribeUserToSubreddit = async (userId: UUID, subreddit: Subreddit) => {
  const user = await userService.fetchUser(userId)
  if (!user) {
    throw new ValidationError('User does not exist!')
  }
  await subreddit.$add('user', user)
};

export const createOrFetchSubreddit = async (name: string) => {
  return withTransaction(async transaction => {
    const subreddit = await fetchSubreddit(name, transaction)
    if (subreddit) {
      return subreddit
    }
    return Subreddit.create({name}, {transaction});
  })
}