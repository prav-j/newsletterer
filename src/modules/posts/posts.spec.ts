import { getTopPostsInLastDay } from "./service";
import Subreddit from "../subreddit/Subreddit.model";
import { resetDB } from "../../db";

jest.mock('node-fetch')

describe('Fetch posts', () => {
  beforeEach(resetDB)
  afterAll(resetDB)

  it('should fetch top posts', async () => {
    const posts = await getTopPostsInLastDay(Subreddit.build({name: 'test'}))
    expect(posts).toMatchSnapshot()
  })
})