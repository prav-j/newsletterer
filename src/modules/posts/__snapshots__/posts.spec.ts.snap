// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fetch posts should fetch top posts 1`] = `
Object {
  "link": "https://www.reddit.com/r/test",
  "logo": "https://b.thumbs.redditmedia.com/slwYpQeWMVyRwpV_cgthYvlk4f2QU2SD5N_Id-dlWmA.png",
  "posts": Array [
    Object {
      "link": "https://www.reddit.com/r/factorio/comments/jyu4yc/til_to_not_store_grenades_in_chests/",
      "thumbnail": "https://b.thumbs.redditmedia.com/e17giyxleTTFcNJPEZTcf374mNuh-sYmN86ceBwbcDk.jpg",
      "title": "TIL to not store grenades in chests",
    },
    Object {
      "link": "https://www.reddit.com/r/factorio/comments/jyi8yg/nopipe_november/",
      "thumbnail": undefined,
      "title": "Nopipe November",
    },
    Object {
      "link": "https://www.reddit.com/r/factorio/comments/jyhoyd/i_can_see_my_factory_breathe_on_the_edge_of_my/",
      "thumbnail": "https://a.thumbs.redditmedia.com/N-1edZc7dJKUVPGRV9iH2-81OLkx7PBlnX_xav207m8.jpg",
      "title": "I can see my factory breathe on the edge of my pollution cloud. Soon it will gain consciousness.",
    },
  ],
  "subtitle": "Subreddit for the game Factorio made by Wube Software.",
  "title": "top scoring links : factorio",
}
`;
